### PR TITLE
add inscopix imaging extractor

### PR DESCRIPTION
**Pull request by bendichter**

### CODE REVIEW

File: .github/workflows/run-tests.yml
Comment (CodyCBakerPhD) [inline .github/workflows/run-tests.yml:24]: Why the strong drop of 3.8 10 months before Python stops supporting it?

Even if `isx` doesn't support it, you can have conditional requirements that say that `roiextractors` + `isx` support requires `py>3.8`, but the rest of `roiextractors` is cool with it

Comment (bendichter) [reply .github/workflows/run-tests.yml:24]: Let's follow [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html), not the [Python release cycle](https://devguide.python.org/versions/). According to NEP 29, Python 3.8 was dropped Apr 14, 2023.

Comment (CodyCBakerPhD) [reply .github/workflows/run-tests.yml:24]: We've been following the Python release cycle everywhere else - maybe discuss in the next team meeting?

Comment (bendichter) [reply .github/workflows/run-tests.yml:24]: The latest version of numpy, 1.26.0, requires Python 3.9+. If we want to use any new numpy features (or any features that rely on them) we will need to drop support for Python 3.8.

Comment (CodyCBakerPhD) [reply .github/workflows/run-tests.yml:24]: What new features would those be, and why would ROIExtractors, NeuroConv, or the GUIDE minimally require them instead of simply letting it be up to the user and their condition installations?

You can also specify `numpy>=1.26.0` conditionally for various Python versions and optional requirements without making the strong decision to retire an entire Python version

Comment (bendichter) [reply .github/workflows/run-tests.yml:24]: > This NEP is being put forward by maintainers of Matplotlib, scikit-learn, IPython, Jupyter, yt, SciPy, NumPy, and scikit-image.

Which means that the latest version of all of those libraries is intended to be 3.9+. See for example https://pypi.org/project/scikit-learn/

Comment (bendichter) [reply .github/workflows/run-tests.yml:24]: This has been put forth by all of those projects as a best practice for ecosystem support. I don't really want to create any unnecessarily packaging complexity to extend beyond this window

Comment (h-mayorquin) [reply .github/workflows/run-tests.yml:24]: @bendichter 
I was not aware of NEP 29. Nice.

I also think that there are minimal costs to dropping support of python 3.8. I think our user base has a lot of freedom over their environments as conda is available on most scientific contexts. As Ben said, I think the cost of increasing packacing complexity is a greater burden for us. If some user comes forward with an issue maybe we can carve out a solution for them but I think it would be easier for us to stick to the scientific python standards.

**Comment by bendichter** [comment]
blocked by https://github.com/inscopix/py_isx/issues/48

**Comment by pauladkisson** [comment]
Looks like this is unblocked as of v1.0.4 (https://github.com/inscopix/py_isx/issues/48#issuecomment-2033057806)

File: requirements-full.txt
Comment (CodyCBakerPhD) [inline requirements-full.txt:5]:
```suggestion
isx>=1.0.4
```

File: src/roiextractors/extractors/inscopixextractors/inscopiximagingextractor.py
Comment (CodyCBakerPhD) [inline src/roiextractors/extractors/inscopixextractors/inscopiximagingextractor.py:3]:
```suggestion
"""Inscopix Imaging Extractor."""
```

File: src/roiextractors/extractors/inscopixextractors/__init__.py
Comment (CodyCBakerPhD) [inline src/roiextractors/extractors/inscopixextractors/__init__.py:3]:
```suggestion
"""Defines extactors for Inscopix data."""
```

File: src/roiextractors/extractors/inscopixextractors/inscopiximagingextractor.py
Comment (CodyCBakerPhD) [inline src/roiextractors/extractors/inscopixextractors/inscopiximagingextractor.py:23]:
```suggestion
        """
        Create an InscopixImagingExtractor instance from a single .isx file.

        Parameters
        ----------
        file_path : PathType
            Path to the Inscopix file.
        """
```

**Comment by CodyCBakerPhD** [comment]
Looks like the test cases just need to be updated now - also a CHANGELOG entry

File: CHANGELOG.md
Comment (CodyCBakerPhD) [inline CHANGELOG.md:5]:
```suggestion
* Add InscopixImagingExtractor [#276](https://github.com/catalystneuro/roiextractors/pull/276)
```

**Comment by bendichter** [comment]
This is finally working. Though I think dropping 3.8 should probably be a different PR

**Comment by CodyCBakerPhD** [comment]
I'll defer to @pauladkisson for PR review but mostly looks good to me

> Though I think dropping 3.8 should probably be a different PR

For the global drop of 3.8 I would agree, and then we can do it synchronously throughout the ecosystem across the next few weeks (I think we have a couple of months still)

However, you would still need various skips of `isx` stuff for 3.8 (such as in the CI installation step too) for this PR so maybe too much bother here?

Could do the 3.8 drop in separate PR, give that PR precedence, then come back to this?

**Comment by pauladkisson** [comment]
Happy to review, but will have to wait until 5/20

**Comment by h-mayorquin** [comment]
https://github.com/catalystneuro/neuroconv/pull/853

**Comment by bendichter** [comment]
depends on https://github.com/catalystneuro/roiextractors/pull/325

File: tests/test_inscopiximagingextractor.py
Comment (h-mayorquin) [inline tests/test_inscopiximagingextractor.py:20]: This is geting the method from the base class which makes a call to a read operation to get the dtype. 

To avoid this read operation everytime that someeone calls dtype maybe we can determine this at `__init__`

I was hoping that this information was available on the footer and there is a dtype but I am not sure how to map it to numpy dtypes. Check out the footer for this test:

```
print(extractor.movie.footer)
{'dataType': 1, 'extraProperties': None, 'fileVersion': 1, 'hasFrameHeaderFooter': False, 'producer': {'name': 'isxcore', 'version': [1, 8, 0]}, 'spacingInfo': {'numPixels': {'x': 128, 'y': 128}, 'pixelSize': {'x': {'den': 1, 'num': 3}, 'y': {'den': 1, 'num': 3}}, 'topLeft': {'x': {'den': 1, 'num': 0}, 'y': {'den': 1, 'num': 0}}}, 'timingInfo': {'blank': [], 'cropped': [], 'dropped': [], 'numTimes': 100, 'period': {'den': 1000, 'num': 100}, 'start': {'secsSinceEpoch': {'den': 1, 'num': 0}, 'utcOffset': 0}}, 'type': 0}
```

**Review by h-mayorquin — COMMENTED** [review]
This looks good, my only request is that we avoid the read operation every time dtype is requested.

File: tests/test_inscopiximagingextractor.py
Comment (h-mayorquin) [reply tests/test_inscopiximagingextractor.py:20]: Ah, this can be extracted from `self.movie.data_type` see:

https://inscopix.github.io/py_isx/how-to/read-movie-data-type.html

File: tests/test_inscopiximagingextractor.py
Comment (bendichter) [reply tests/test_inscopiximagingextractor.py:41]: I'm on the fence. I was keeping them on in case users run 3.8 tests. On the other hand, maybe it's best for them to fail to inform the user that inscopix is not supported on their system

**Comment by bendichter** [comment]
@h-mayorquin I added a method as you suggested. I also added a test to ensure that the dtype of the data indeed matches this dtype extacted from the metadata (this did not match in a previous version of the isx package)

**Review by h-mayorquin — APPROVED** [review]
All right.

**Pull request merged**
